### ISSUE TITLE
[bitnami/harbor] registry pod configuration error: error parsing /etc/registry/config.yml 

### ISSUE THREAD

**Issue by luozhihua** [bitnami/charts#1954]
**Which chart**:
  harbor v4.0.0
  repository: https://hub.helm.sh/charts/bitnami/harbor

**Description**

harbor registry pod not started successfully, logs of harbor registry pod:
```bash
$> kubectl log -n registry pod/harbor-registry-66c8dd44b7-58qhj -c registry

log is DEPRECATED and will be removed in a future version. Use logs instead.
 03:45:22.87
 03:45:22.87 Welcome to the Bitnami harbor-registry container
 03:45:22.87 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-harbor-registry
 03:45:22.87 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-harbor-registry/issues
 03:45:22.87 Send us your feedback at containers@bitnami.com
 03:45:22.88
 03:45:22.88 INFO  ==> ** Starting Harbor Registry setup **
 03:45:22.89 INFO  ==> Validating Harbor Registry settings...
 03:45:22.89 INFO  ==> ** Harbor Registry setup finished! **
 03:45:22.89 INFO  ==> ** Starting Harbor Registry **
configuration error: error parsing /etc/registry/config.yml: yaml: found character that cannot start any token

Usage:
  registry serve <config> [flags]
Flags:
  -h, --help=false: help for serve

Additional help topics:
```

**Steps to reproduce the issue:**
1. create custom chart `./charts/registry` and set `harbor@v4.0.0` as `dependencies`.

chart.yaml:
```yaml
apiVersion: v2
name: registry
description: A Helm chart for Kubernetes
type: application
version: 0.1.0
appVersion: 1.16.0
dependencies:
  - name: nfs-client-provisioner
    version: 1.2.8
    repository: "@stable" # https://hub.helm.sh/charts/stable/nfs-client-provisioner
  - name: harbor
    version: 4.0.0
    repository: "@bitnami"
```

value.yaml:
```yaml
global:
  storageClass: &storageClass nfs-registry

# =======================================================
# NFS Storage
# https://hub.helm.sh/charts/stable/nfs-client-provisioner
# =======================================================
nfs-client-provisioner:
  replicaCount: 1
  fullnameOverride: "registry"
  image:
    tag: v3.1.0-k8s1.11
  storageClass:
    name: *storageClass
    reclaimPolicy: Retain
    archiveOnDelete: false
  nfs:
    server: 192.168.31.100
    path: /nfs/dev.registry

# =======================================================
# Harbor
# https://hub.helm.sh/charts/bitnami/harbor
# =======================================================
harbor:
  secret: &secret xxxxxxxxxxxxxxxx
  global:
    storageClass: *storageClass
  fullnameOverride: harbor
  externalURL: https://registry.fengwo.io
  forcePassword: true
  harborAdminPassword: xxxxxxxx
  secretKey: harbor-admin-password
  persistence:
    enabled: true
  service:
    type: ClusterIP
    tls:
      enabled: false
  nginx:
    replicas: 1
    behindReverseProxy: true
  portal:
    replicas: 1
  core:
    replicas: 1
    secret: *secret
  Jobservice:
    replicas: 1
    secret: *secret
  registry:
    replicas: 1
    secret: *secret
  registryImage:
    debug: true
  chartmuseum:
    replicas: 1
  clair:
    replicas: 1
  postgresql:
    enabled: false
  redis:
    enabled: false
  externalDatabase:
    host: postgresql.database
    user: postgres
    password: 'xxxxxxxx'
    sslmode: disable
    database: harbor
    coreDatabase: harbor_core
    clairDatabase: harbor_clair
    notaryServerDatabase: harbor_notary_server
    notarySignerDatabase: harbor_notary_signer
  externalRedis:
    host: redis-master.database
    password: 'xxxxxxxx'

# To disable postgres & redis of harbor chart which in dependencies.
postgresql:
  enabled: false
redis:
  enabled: false
```
2. Deploy chart (install or upgrade):

```bash
$> helm upgrade --install registry -n registry ./registry

Release "registry" has been upgraded. Happy Helming!
NAME: registry
LAST DEPLOYED: Thu Feb 20 12:44:08 2020
NAMESPACE: registry
STATUS: deployed
REVISION: 20
TEST SUITE: None
```
3. Check status of the deployment via `kubectl`:
```bash
$> kubectl get pod -n registry

NAME                                    READY   STATUS             RESTARTS   AGE
harbor-chartmuseum-566f6bb958-45jmp     1/1     Running            0          11h
harbor-clair-69b86f87f8-zn9b2           2/2     Running            0          11h
harbor-core-7b7cf9c58f-z46n8            1/1     Running            0          64s
harbor-jobservice-f7c8fc6dc-dlsmk       1/1     Running            0          64s
harbor-nginx-787c6476fd-zq68v           1/1     Running            3          11h
harbor-notary-server-764567c664-8db6g   1/1     Running            0          63s
harbor-notary-signer-74d445bdd7-8tw2l   1/1     Running            0          63s
harbor-portal-7f8f74d669-dkmx6          1/1     Running            0          11h
harbor-registry-66c8dd44b7-58qhj        0/2     CrashLoopBackOff   6          63s
harbor-registry-d497c97bc-6x2nw         1/2     CrashLoopBackOff   16         60m
registry-597c468898-grz4n               1/1     Running            0          11h
```

**Describe the results you received:**

Check logs of pod `harbor-registry-xxx`:
```bash
$> kubectl log -n registry pod/harbor-registry-66c8dd44b7-58qhj -c registry

log is DEPRECATED and will be removed in a future version. Use logs instead.
 03:45:22.87
 03:45:22.87 Welcome to the Bitnami harbor-registry container
 03:45:22.87 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-harbor-registry
 03:45:22.87 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-harbor-registry/issues
 03:45:22.87 Send us your feedback at containers@bitnami.com
 03:45:22.88
 03:45:22.88 INFO  ==> ** Starting Harbor Registry setup **
 03:45:22.89 INFO  ==> Validating Harbor Registry settings...
 03:45:22.89 INFO  ==> ** Harbor Registry setup finished! **
 03:45:22.89 INFO  ==> ** Starting Harbor Registry **
configuration error: error parsing /etc/registry/config.yml: yaml: found character that cannot start any token

Usage:
  registry serve <config> [flags]
Flags:
  -h, --help=false: help for serve

Additional help topics:
```

**Describe the results you expected:**
Expect the registry to start successfully.

**Additional information you deem important (e.g. issue happens only occasionally):**

Partial output of `helm upgrade --install registry -n registry ./registry --debug --dry-run`:

```yaml
# ConfigMap
---
# Source: registry/charts/harbor/templates/registry/registry-cm.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: harbor-registry
  labels:
    app.kubernetes.io/name: "harbor"
    helm.sh/chart: "harbor-4.0.0"
    app.kubernetes.io/instance: registry
    app.kubernetes.io/managed-by: Helm
data:
  config.yml: |+
    version: 0.1
    log:
      level: debug
      fields:
        service: registry
    storage:
      filesystem:
        rootdirectory: /storage
      cache:
        layerinfo: redis
      maintenance:
        uploadpurging:
          enabled: false
      delete:
        enabled: true
      redirect:
        disable: false
    redis:
      addr: "redis-master.database:6379"
      db: 2
    http:
      addr: "localhost:5000"
      # set via environment variable
      # secret: placeholder
      debug:
        addr: "localhost:5001"
    auth:
      token:
        issuer: harbor-token-issuer
        realm: "https://registry.fengwo.io/service/token"
        rootcertbundle: "/etc/registry/root.crt"
        service: harbor-registry
    validation:
      disabled: true
    notifications:
      endpoints:
        - name: harbor
          disabled: false
          url: "http://harbor-core/service/notifications"
          timeout: 3000ms
          threshold: 5
          backoff: 1s
  ctl-config.yml: |+
    ---
    protocol: "http"
    port: 8080
    log_level: debug

# Deployment
---
# Source: registry/charts/harbor/templates/registry/registry-dpl.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: harbor-registry
  labels:
    app.kubernetes.io/name: "harbor"
    helm.sh/chart: "harbor-4.0.0"
    app.kubernetes.io/instance: registry
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/component: registry
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/instance: registry
      app.kubernetes.io/name: "harbor"
      app.kubernetes.io/component: registry
  template:
    metadata:
      labels:
        app.kubernetes.io/name: "harbor"
        helm.sh/chart: "harbor-4.0.0"
        app.kubernetes.io/instance: registry
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/component: registry
      annotations:
        checksum/configmap: 78a57ddf4570bdb7b1d548e617fbbb986254aefe24410d4d655fd1fb94793091
        checksum/secret: a9bca2a50614f056492147f1e173857bfb7bb08e48e612ed4a51cd31d3811a23
        checksum/secret-jobservice: 76c0322041ccee31e53bd472fcdf78e70b2434d6872dee599309cc4201b7f77b
        checksum/secret-core: d244c527346be6ad0a86794fc13f485f86bf9d932de4ca0a02250d05e9261326
    spec:      
      securityContext:
        fsGroup: 1001
        runAsUser: 1001
      containers:
        - name: registry
          image: docker.io/bitnami/harbor-registry:1.10.1-debian-10-r0
          imagePullPolicy: "IfNotPresent"
          livenessProbe:
            httpGet:
              path: /
              port: registry
            initialDelaySeconds: 10
            periodSeconds: 10
            timeoutSeconds: 5
            successThreshold: 1
            failureThreshold: 6
          readinessProbe:
            httpGet:
              path: /
              port: registry
            initialDelaySeconds: 10
            periodSeconds: 10
            timeoutSeconds: 5
            successThreshold: 1
            failureThreshold: 6
          resources:
            limits: {}
            requests: {}
          env:
            - name: BITNAMI_DEBUG
              value: "true"
          envFrom:
            - secretRef:
                name: harbor-registry
          ports:
            - containerPort: 5000
              name: registry
            - containerPort: 5001
              name: debug
          volumeMounts:
            - name: registry-data
              mountPath: /storage
              subPath: 
            - name: registry-root-certificate
              mountPath: /etc/registry/root.crt
              subPath: tls.crt
            - name: registry-config
              mountPath: /etc/registry/config.yml
              subPath: config.yml
        - name: registryctl
          image: docker.io/bitnami/harbor-registryctl:1.10.1-debian-10-r0
          imagePullPolicy: "IfNotPresent"
          livenessProbe:
            httpGet:
              path: /api/health
              port: registryctl
            initialDelaySeconds: 10
            periodSeconds: 10
            timeoutSeconds: 5
            successThreshold: 1
            failureThreshold: 6
          readinessProbe:
            httpGet:
              path: /api/health
              port: registryctl
            initialDelaySeconds: 10
            periodSeconds: 10
            timeoutSeconds: 5
            successThreshold: 1
            failureThreshold: 6
          resources:
            limits: {}
            requests: {}
          envFrom:
            - secretRef:
                name: harbor-registry
          env:
            - name: BITNAMI_DEBUG
              value: "false"
            - name: CORE_SECRET
              valueFrom:
                secretKeyRef:
                  name: harbor-core
                  key: secret
            - name: JOBSERVICE_SECRET
              valueFrom:
                secretKeyRef:
                  name: harbor-jobservice
                  key: secret
          ports:
            - containerPort: 8080
              name: registryctl
          volumeMounts:
            - name: registry-data
              mountPath: /storage
              subPath: 
            - name: registry-config
              mountPath: /etc/registry/config.yml
              subPath: config.yml
            - name: registry-config
              mountPath: /etc/registryctl/config.yml
              subPath: ctl-config.yml
      volumes:
        - name: registry-root-certificate
          secret:
            secretName: harbor-core
        - name: registry-config
          configMap:
            name: harbor-registry
        - name: registry-data
          persistentVolumeClaim:
            claimName: harbor-registry
---

```

**Version of Helm and Kubernetes**:

- Output of `helm version` (formatted): 

```javascript
version.BuildInfo{
  Version:"v3.0.3", 
  GitCommit:"ac925eb7279f4a6955df663a0128044a8a6b7593", 
  GitTreeState:"clean", 
  GoVersion:"go1.13.7"
}
```

- Output of `kubectl version` (formatted):

```javascript
Client Version: version.Info {
  Major:"1", 
  Minor:"15", 
  GitVersion:"v1.15.5", 
  GitCommit:"20c265fef0741dd71a66480e35bd69f18351daea", 
  GitTreeState:"clean", 
  BuildDate:"2019-10-15T19:16:51Z", 
  GoVersion:"go1.12.10", 
  Compiler:"gc", 
  Platform:"darwin/amd64"
}
Server Version: version.Info {
  Major:"1", 
  Minor:"17", 
  GitVersion:"v1.17.2", 
  GitCommit:"59603c6e503c87169aea6106f57b9f242f64df89",
  GitTreeState:"clean", 
  BuildDate:"2020-01-18T23:22:30Z", 
  GoVersion:"go1.13.5", 
  Compiler:"gc", 
  Platform:"linux/amd64"
}
```

**Comment by javsalgar** [comment]
Hi,

I was unable to reproduce the issue. I didn't use the nfs provisioner but the default storageclass and registry worked without issues

```
 130 ❯ stern -n registry harbor-registry-b64b7bd74-5lkrh
+ harbor-registry-b64b7bd74-5lkrh › registryctl
+ harbor-registry-b64b7bd74-5lkrh › registry
harbor-registry-b64b7bd74-5lkrh registry  10:16:59.00
harbor-registry-b64b7bd74-5lkrh registry  10:16:59.00 Welcome to the Bitnami harbor-registry container
harbor-registry-b64b7bd74-5lkrh registry  10:16:59.01 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-harbor-registry
harbor-registry-b64b7bd74-5lkrh registry  10:16:59.01 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-harbor-registry/issues
harbor-registry-b64b7bd74-5lkrh registry  10:16:59.02 Send us your feedback at containers@bitnami.com
harbor-registry-b64b7bd74-5lkrh registry  10:16:59.02
harbor-registry-b64b7bd74-5lkrh registry  10:16:59.02 INFO  ==> ** Starting Harbor Registry setup **
harbor-registry-b64b7bd74-5lkrh registry  10:16:59.03 INFO  ==> Validating Harbor Registry settings...
harbor-registry-b64b7bd74-5lkrh registry
harbor-registry-b64b7bd74-5lkrh registry  10:16:59.03 INFO  ==> ** Harbor Registry setup finished! **
harbor-registry-b64b7bd74-5lkrh registry  10:16:59.03 INFO  ==> ** Starting Harbor Registry **
harbor-registry-b64b7bd74-5lkrh registry time="2020-02-20T10:16:59.060594212Z" level=info msg="debug server listening localhost:5001"
harbor-registry-b64b7bd74-5lkrh registry time="2020-02-20T10:16:59.063032943Z" level=info msg="configuring endpoint harbor (http://harbor-core/service/notifications), timeout=3s, headers=map[]" go.version=go1.13.8 instance.id=7350ad01-d1bd-4e23-ad9c-bd3aea0f5e64 service=registry version=v2.7.1
harbor-registry-b64b7bd74-5lkrh registry time="2020-02-20T10:16:59.075075201Z" level=info msg="using redis blob descriptor cache" go.version=go1.13.8 instance.id=7350ad01-d1bd-4e23-ad9c-bd3aea0f5e64 service=registry version=v2.7.1
harbor-registry-b64b7bd74-5lkrh registry time="2020-02-20T10:16:59.075390432Z" level=debug msg="configured "token" access controller" go.version=go1.13.8 instance.id=7350ad01-d1bd-4e23-ad9c-bd3aea0f5e64 service=registry version=v2.7.1
harbor-registry-b64b7bd74-5lkrh registry time="2020-02-20T10:16:59.075542638Z" level=info msg="listening on [::]:5000" go.version=go1.13.8 instance.id=7350ad01-d1bd-4e23-ad9c-bd3aea0f5e64 service=registry version=v2.7.1
harbor-registry-b64b7bd74-5lkrh registry 10.32.0.1 - - [20/Feb/2020:10:17:09 +0000] "GET / HTTP/1.1" 200 0 "" "kube-probe/1.17"
harbor-registry-b64b7bd74-5lkrh registry 10.32.0.1 - - [20/Feb/2020:10:17:16 +0000] "GET / HTTP/1.1" 200 0 "" "kube-probe/1.17"
harbor-registry-b64b7bd74-5lkrh registry 10.32.0.1 - - [20/Feb/2020:10:17:19 +0000] "GET / HTTP/1.1" 200 0 "" "kube-probe/1.17"
harbor-registry-b64b7bd74-5lkrh registry 10.32.0.1 - - [20/Feb/2020:10:17:26 +0000] "GET / HTTP/1.1" 200 0 "" "kube-probe/1.17"
harbor-registry-b64b7bd74-5lkrh registry 10.32.0.1 - - [20/Feb/2020:10:17:29 +0000] "GET / HTTP/1.1" 200 0 "" "kube-probe/1.17"
harbor-registry-b64b7bd74-5lkrh registry 10.32.0.1 - - [20/Feb/2020:10:17:36 +0000] "GET / HTTP/1.1" 200 0 "" "kube-probe/1.17"
harbor-registry-b64b7bd74-5lkrh registry 10.32.0.1 - - [20/Feb/2020:10:17:39 +0000] "GET / HTTP/1.1" 200 0 "" "kube-probe/1.17"
harbor-registry-b64b7bd74-5lkrh registry 10.32.0.1 - - [20/Feb/2020:10:17:46 +0000] "GET / HTTP/1.1" 200 0 "" "kube-probe/1.17"
harbor-registry-b64b7bd74-5lkrh registry 10.32.0.1 - - [20/Feb/2020:10:17:49 +0000] "GET / HTTP/1.1" 200 0 "" "kube-probe/1.17"
harbor-registry-b64b7bd74-5lkrh registry 10.32.0.1 - - [20/Feb/2020:10:17:56 +0000] "GET / HTTP/1.1" 200 0 "" "kube-probe/1.17"
```


Generated configmap

```
❯ kubectl get cm -n registry -o yaml harbor-registry
apiVersion: v1
data:
  config.yml: |
    version: 0.1
    log:
      level: debug
      fields:
        service: registry
    storage:
      filesystem:
        rootdirectory: /storage
      cache:
        layerinfo: redis
      maintenance:
        uploadpurging:
          enabled: false
      delete:
        enabled: true
      redirect:
        disable: false
    redis:
      addr: "redis-master.database:6379"
      db: 2
    http:
      addr: :5000
      # set via environment variable
      # secret: placeholder
      debug:
        addr: localhost:5001
    auth:
      token:
        issuer: harbor-token-issuer
        realm: "https://registry.fengwo.io/service/token"
        rootcertbundle: /etc/registry/root.crt
        service: harbor-registry
    validation:
      disabled: true
    notifications:
      endpoints:
        - name: harbor
          disabled: false
          url: http://harbor-core/service/notifications
          timeout: 3000ms
          threshold: 5
          backoff: 1s
  ctl-config.yml: |
    ---
    protocol: "http"
    port: 8080
    log_level: debug
kind: ConfigMap
metadata:
  creationTimestamp: "2020-02-20T10:16:42Z"
  labels:
    app.kubernetes.io/instance: registry
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: harbor
    helm.sh/chart: harbor-4.0.0
  name: harbor-registry
  namespace: registry
  resourceVersion: "12442"
  selfLink: /api/v1/namespaces/registry/configmaps/harbor-registry
  uid: 82efd958-cf83-4666-9f13-694d1cf9f0c0
```

**Comment by luozhihua** [comment]
@javsalgar  thank you very much for your precious time to help me troubleshoot. I found the reason now, because I used a password containing "`@`" character in a secrets, which caused various strange template syntax problems, so, This has nothing to do with chart `bitnami/harbor`.

In the past two days, I spent a lot of energy to troubleshoot this inexplicable problem. I almost checked each character of the `config-map` and `deployment` files one by one, and finally compared the correct test results you provided. The configuration given to you is different from the password, only to discover the real reason, thank you again!
